File: API/Entities/Product.cs
using System.ComponentModel.DataAnnotations;
using API.Entities.enums;

namespace API.Entities
{
    public class Product
    {
        [Key]
        public int ProductId { get; set; }

        public string? Name { get; set; }

        public string? Description { get; set; }

        public long Price { get; set; }

        public string? PictureUrl { get; set; }

        public int QuantityInStock { get; set; }

        public Category? Category { get; set; }
        public int CategoryId { get; set; }

        // Product Type - Physical or Digital
        public ProductType ProductType { get; set; } = ProductType.Physical;

        // Digital Product specific fields
        public string? DigitalFileUrl { get; set; }  // URL for digital download
        public bool IsInstantDelivery { get; set; } = false;  // Auto-deliver after payment

        public string? PublicId { get; set; }

        public ICollection<ProductVariant> Variants { get; set; } = new List<ProductVariant>();
        public void AddVariant(List<int> attributeValueIds, int quantity)
        {
            // Proveri da li već postoji variant sa istim atributima
            var existingVariant = Variants.FirstOrDefault(v =>
                v.Attributes.Count == attributeValueIds.Count &&
                v.Attributes.All(attr => attributeValueIds.Contains(attr.AttributeValueId))
            );

            if (existingVariant == null)
            {
                // Kreiraj novi variant
                var newVariant = new ProductVariant
                {
                    ProductId = this.ProductId,
                    QuantityInStock = quantity,
                    Attributes = attributeValueIds.Select(id => new ProductVariantAttribute
                    {
                        AttributeValueId = id
                    }).ToList()
                };

                Variants.Add(newVariant);
            }
            else
            {
                // Ako već postoji, samo poveća količinu
                existingVariant.QuantityInStock += quantity;
            }
        }

    }
}

File: client/src/app/models/product.ts
export enum ProductType {
  Physical = 0,
  Digital = 1
}

export interface Product {
    productId: number
    name: string
    price: number
    description: string
    pictureUrl: string
    imageUrl?: string  // Backend returns imageUrl
    categoryName: string
    categoryId: number
    productType: ProductType
    digitalFileUrl?: string
    isInstantDelivery: boolean
    variants?: ProductVariant[]
  }
  
  export interface Category {
    categoryId: number
    name: string
  }

  export interface ProductVariant {
  productVariantId: number
  productId: number
  price: number               // Actual price to display (priceOverride if set, otherwise base price)
  priceOverride?: number | null  // Custom price for this variant (null if using base price)
  quantityInStock: number
  attributeValueIds: number[]  // IDs od izabranih vrednosti atributa
  attributes?: VariantAttribute[] // detaljnije info o atributima
}

export interface VariantAttribute {
  attributeName: string
  attributeValue: string
}
  export interface ProductParams {
    orderBy: string;
    searchTerm?: string;
    categories: string[];
    attributeValueIds?: number[];
    productTypes: string[];
    pageNumber: number;
    pageSize: number;
}
